### PR TITLE
feat: Istio module addon

### LINKED ISSUE #1107: Add Istio module add-on
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

Terraform module for deploying [Istio](https://istio.io/) via Helm.

#### Describe the solution you would like

Include the istio addon in the `kubernetes-addons` section of this blueprint.

### PR DESCRIPTION
**Pull request by ovaleanu**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

Adds istio addon. This addon allows to deploy [Istio](https://istio.io/), an open source service mesh that layers transparently onto existing distributed applications.

### Motivation

- Resolves #1107 

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs require a new example and/or doc page. In general:
- Use an existing example when possible to demonstrate a new addons usage
- A new docs page under `docs/add-ons/*` is required for new a new addon

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Review by askulkarni2 — COMMENTED** [review]
@ovaleanu thanks for the PR. Looking good, can you please resolve merge conflicts?

File: modules/kubernetes-addons/istio/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<!-- BEGIN_TF_DOCS -->
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/README.md:1]: please see the other readmes for the correct tags

File: modules/kubernetes-addons/istio/main.tf [hunk 2]
```diff
@@ -0,0 +1,138 @@
+resource "kubernetes_namespace" "istio_system" {
+  metadata {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:2]: namespace creation and name used should be configurable - see other examples for reference

File: modules/kubernetes-addons/istio/main.tf [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    name = "istio-system"
+    labels = {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:4]: is this required? are there other labels that we need to allow users to pass in?

File: modules/kubernetes-addons/istio/main.tf [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        yamlencode(var.istio_base_settings)
+      ]
+      description = "Helm chart for deploying Istio cluster resources and CRDs"
+    },
+    var.helm_config
+  )
+  manage_via_gitops = var.manage_via_gitops
+  addon_context     = var.addon_context
+
+  depends_on = [kubernetes_namespace.istio_system]
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:34]: these `depends_on` are not required due to implicit dependency shown above

File: modules/kubernetes-addons/istio/main.tf [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  count  = var.install_istio_base ? 1 : 0
+
+  helm_config = merge(
+    {
+      name            = "istio-base"
+      chart           = "base"
+      repository      = "https://istio-release.storage.googleapis.com/charts"
+      version         = var.istio_version
+      namespace       = kubernetes_namespace.istio_system.metadata.0.name
+      timeout         = 120
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:21]: we can remove all of these from each one
- `timeout`
- `cleanup_on_fail`
- `force_update`
- `values`

unless there is a strong reason to deviate from the defaults

File: modules/kubernetes-addons/istio/main.tf [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+module "istio_base" {
+  source = "../helm-addon"
+  count  = var.install_istio_base ? 1 : 0
+
+  helm_config = merge(
+    {
+      name            = "istio-base"
+      chart           = "base"
+      repository      = "https://istio-release.storage.googleapis.com/charts"
+      version         = var.istio_version
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:19]: set version here not as a top level variable

File: modules/kubernetes-addons/istio/versions.tf [hunk 7]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0.1"
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/versions.tf:2]:
```suggestion
  required_version = ">= 1.0.0"
```

File: modules/kubernetes-addons/variables.tf [hunk 8]
```diff
@@ -265,6 +265,43 @@ variable "crossplane_kubernetes_provider" {
   }
 }
 
+#-----------ISTIO ADDON------------------
+variable "enable_istio" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/variables.tf:269]: currently this doesn't do anything if changed to `true`. what is the "default" for installing Istio - this should be the behavior that is enabled when setting this to `true`

should base, cni, and istiod all be installed in the "default" scenario? or is only "base" required for the MVP of Istio install?

File: modules/kubernetes-addons/variables.tf [hunk 9]
```diff
@@ -265,6 +265,43 @@ variable "crossplane_kubernetes_provider" {
...
+  default     = false
+}
+
+variable "istio_version" {
+  description = "Istio version"
+  type        = string
+  default     = ""
+}
+
+variable "install_istio_base" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/variables.tf:281]: we use `enable` by convention

File: modules/kubernetes-addons/istio/main.tf [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+      namespace       = kubernetes_namespace.istio_system.metadata.0.name
+      timeout         = 120
+      cleanup_on_fail = var.cleanup_on_fail
+      force_update    = var.force_update
+      values = [
+        yamlencode(var.istio_base_settings)
+      ]
+      description = "Helm chart for deploying Istio cluster resources and CRDs"
+    },
+    var.helm_config
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:29]: each one of these helm_releases need their own set of configurations. See the Kyverno addon for reference

File: modules/kubernetes-addons/istio/main.tf [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    {
+      name            = "istio-cni"
+      chart           = "cni"
+      repository      = "https://istio-release.storage.googleapis.com/charts"
+      version         = var.istio_version
+      namespace       = kubernetes_namespace.istio_system.metadata.0.name
+      timeout         = 120
+      cleanup_on_fail = var.cleanup_on_fail
+      force_update    = var.force_update
+      values = [
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:51]: is this value required to function correctly? if not, don't set it as a default and let users pass in their own values

File: modules/kubernetes-addons/istio/main.tf [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    {
+      name            = "istiod"
+      chart           = "istiod"
+      repository      = "https://istio-release.storage.googleapis.com/charts"
+      version         = var.istio_version
+      namespace       = kubernetes_namespace.istio_system.metadata.0.name
+      timeout         = 120
+      cleanup_on_fail = var.cleanup_on_fail
+      force_update    = var.force_update
+      values = [
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:84]: same here - only set required defaults

File: modules/kubernetes-addons/istio/main.tf [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        )
+      ]
+      description = "Helm chart for istio control plane"
+    },
+    var.helm_config
+  )
+  manage_via_gitops = var.manage_via_gitops
+  addon_context     = var.addon_context
+
+  depends_on = [module.istio_cni]
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:110]: for an dependencies - pass an output from the parent to the child to map dependency hierarchy. we do not want to use `depends_on` across modules, its very disruptive and will cause outages in this context

**Review by bryantbiggs — CHANGES_REQUESTED** [review]

File: docs/add-ons/istio.md [hunk 14]
```diff
@@ -0,0 +1,33 @@
+# Istio
+
+[Istio](https://istio.io/) is an open source service mesh that layers transparently onto existing distributed applications. Istio’s powerful features provide a uniform and more efficient way to secure, connect, and monitor services.
```
Comment (bryantbiggs) [inline docs/add-ons/istio.md:3]: Lets provide more actionable information - if someone is enabling Istio, they most likely already know what it is so this current statement doesn't provide much value

For example - theres 4 components within this module - whats the relationships, what are the different combinations and why, what does each combination give users, what is the bare minimum MVP case and whats advanced? What about the other charts in the istio repository https://github.com/istio/istio/tree/master/manifests/charts?

File: modules/kubernetes-addons/istio/main.tf [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  source = "../helm-addon"
+  count  = var.install_istio_cni ? 1 : 0
+
+  helm_config = merge(
+    {
+      name            = "istio-cni"
+      chart           = "cni"
+      repository      = "https://istio-release.storage.googleapis.com/charts"
+      version         = var.istio_version
+      namespace       = kubernetes_namespace.istio_system.metadata.0.name
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:47]: this states `CNI must run in a dedicated namespace, it is a 'singleton' and extremely security sensitive. Access to the CNI namespace must be highly restricted` - so should this be in the same namespace as the other components?

File: modules/kubernetes-addons/istio/main.tf [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  depends_on = [module.istio_cni]
+}
+
+module "istio_ingressgateway" {
+  source = "../helm-addon"
+  count  = var.install_istio_ingressgateway ? 1 : 0
+
+  helm_config = merge(
+    {
+      name            = "istio-ingressgateway"
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/istio/main.tf:119]: this states `A cluster may use multiple Gateways, each with a different load balancer IP, domains and certificates.` - how do we support this in the configuration currently proposed?

File: modules/kubernetes-addons/istio/main.tf (hunk 2 above)
Comment (ovaleanu) [reply modules/kubernetes-addons/istio/main.tf:2]: Istio control plane can be installed only in `istio-system` namespace

**Comment by bryantbiggs** [comment]
closing for now - please feel free to re-open once you've had a chance to address the requested changes

**Pull request closed without merging**